### ISSUE TITLE
CollectionAssert (mis)handling of IList implementation

### ISSUE THREAD

**Issue by WindingWinter** [nunit/nunit#4635]
I found that somehow, if a class is `IList` and `IList<T>`, then it will use `CopyTo(Array array, int index)`, despite that it could have ( or shall I say, **should** have) use the generic version,  `CopyTo(T[] array, int index)`. Here's how to reproduce the issue.

This is the ListExt

```
 public class ListExt:IList<double>, IList 
 {
   
     private readonly List<double> _inner = new List<double>();

     public IEnumerator<double> GetEnumerator()
     {
         return _inner.GetEnumerator();
     }

     IEnumerator IEnumerable.GetEnumerator()
     {
         return GetEnumerator();
     }

     public void Add(double item)
     {
         _inner.Add(item);
     }



     public int Add(object value)
     {
         Add((double)value);
         return Count - 1; // Return the index where the item was added
     }

     public bool Contains(object value)
     {
         return Contains((double)value);
     }

     void IList.Clear()
     {
         _inner.Clear();
     }
     void IList.RemoveAt(int index)
     {
         _inner.RemoveAt(index);
     }

     object IList.this[int index]
     {
         get => this[index];
         set => this[index] = (double)value;
     }

     bool IList.IsReadOnly => false;

     int ICollection.Count => Count;

     public int IndexOf(object value)
     {
         return IndexOf((double)value);
     }

     public void Insert(int index, object value)
     {
        Insert(index, (double)value);
     }

     public void Remove(object value)
     {
         Remove((double)value);
     }



     public bool IsFixedSize => throw new NotImplementedException();

     void ICollection<double>.Clear()
     {
        _inner.Clear();
     }

     public bool Contains(double item)
     {
         return _inner.Contains(item);
     }

     public void CopyTo(double[] array, int arrayIndex)
     {
         _inner.CopyTo(array, arrayIndex);
     }

     public bool Remove(double item)
     {
      return  _inner.Remove(item);
     }

     public void CopyTo(Array array, int index)
     {
         throw new NotImplementedException();
     }

     

     public object SyncRoot => throw new NotImplementedException();
     public bool IsSynchronized => throw new NotImplementedException();

     int Count => _inner.Count;
     int ICollection<double>.Count => Count;


     bool ICollection<double>.IsReadOnly => false;

     public int IndexOf(double item)
     {
         return _inner.IndexOf(item);
     }

     public void Insert(int index, double item)
     {
         _inner.Insert(index, item);
     }

     void IList<double>.RemoveAt(int index)
     {
        _inner.RemoveAt(index);
     }

     public double this[int index]
     {
         get => _inner[index];
         set => _inner[index]=value;
     }
 }
```

And here's the test case:

```
        [Test]
        public void ListInsertionTest()
        {
            var aList = new ListExt();
            aList.Add(3.0);
            aList.Add(4.0);
            
            
            var bList = new ListExt();
            bList.Add(1.2);
            bList.Add(4.0);
            
            CollectionAssert.AreEquivalent(aList, bList);
        }

```

The test fails with `NotImplementedException` because it calls `CopyTo(Array array, int index)`. I would think that since both the `IList` and its generic version are there, then the NUnit should call the generic version of `CopyTo` for whatever operation. 

Now, if I remove `IList` interface ( and leaves only `IList<double>` interface) and also the methods that implement `IList` explicitly ( to make it compile), then the test will still fail, but now it fails with a more correct error, which is the two lists aren't the same.

**Comment by manfred-brands** [comment]
@WindingWinter A lot of the NUnit code pre-dates generics and most values are internally hold as object.

The signature of the `CollectionAssert` method is:
```csharp
public static void AreEquivalent(IEnumerable expected, IEnumerable actual)
```

We have attempted various times to add the knowledge of `T` to this, but we always fell into a rabbit hole.
But I haven't given up as I would like to have it myself.

Having said that, I cannot find a call to `IList.CopyTo`, not even a call to `IList`.
Can you show me a stacktrace please?

**Comment by WindingWinter** [comment]
@manfred-brands ,

Here's the stack trace

```
System.NotImplementedException : The method or operation is not implemented.
   at NUnitIList.ListExt.CopyTo(Array array, Int32 index) in E:\.net Sample Project\February 2024\NUnitIList\NUnitIList\ListExt.cs:line 103
   at System.Collections.ArrayList.InsertRange(Int32 index, ICollection c)
   at NUnit.Framework.Constraints.CollectionTally.ToArrayList(IEnumerable items)
   at NUnit.Framework.Constraints.CollectionTally..ctor(NUnitEqualityComparer comparer, IEnumerable c)
   at NUnit.Framework.Constraints.CollectionItemsEqualConstraint.Tally(IEnumerable c)
   at NUnit.Framework.Constraints.CollectionEquivalentConstraint.Matches(IEnumerable actual)
   at NUnit.Framework.Constraints.CollectionEquivalentConstraint.ApplyTo[TActual](TActual actual)
   at NUnit.Framework.Assert.That[TActual](TActual actual, IResolveConstraint expression, String message, Object[] args)
   at NUnit.Framework.CollectionAssert.AreEquivalent(IEnumerable expected, IEnumerable actual, String message, Object[] args)
   at NUnit.Framework.CollectionAssert.AreEquivalent(IEnumerable expected, IEnumerable actual)
   at NUnitIList.ListExtTest.ListInsertionTest() in E:\.net Sample Project\February 2024\NUnitIList\NUnitIList\ListExtTest.cs:line 26
```

**Comment by manfred-brands** [comment]
Thanks. I see that this is from NUnit version 3.x where we create an `ArrayList` a standard .NET class also predating generics which then [calls into your implementation of `ICollection`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Collections/ArrayList.cs#L404)

NUnit 4.x  uses `List<object>` internally.

**Comment by stevenaw** [comment]
@WindingWinter Can you please confirm which version of the nunit framework you're using, and if it also occurs in 3.14 (the latest of the 3.x line)?

**Comment by stevenaw** [comment]
Oh I see, that exception is thrown by your code. @manfred-brands is right, internally in nunit 3 favors non-generic approaches in many places.

Based on your stack trace, I suspect you're also hitting https://github.com/nunit/nunit/pull/4097. Can you try your code on 3.14?

**Comment by WindingWinter** [comment]
It occurs in 3.13.x

I haven't test in other version though. Is there any reason to suppose that
this is fixed in the higher version

**Comment by stevenaw** [comment]
3.14 will not call the generic `CopyTo<>()` method, nor will I think we are likely to change that in a future 3.x release. However, that PR _does_ remove a call to the non-generic `CopyTo()` being done within the ArrayList constructor and the change was released in 3.14. If it is the same issue you're hitting then upgrading may save you having to implement the non-generic `CopyTo()` function within your own class.

**Comment by stevenaw** [comment]
I've confirmed this is the same issue as https://github.com/nunit/nunit/pull/4097
Repro: https://github.com/nunit/nunit.issues/tree/main/Issue4635
